Python Classes Module - closes #676 (#680)
bob-crutchley/courseware: topics/python/modules/classes/README.md

	- [Why use classes?](#why-use-classes)
	- [Class and Object example](#class-and-object-example)
	- [The Class Constructor](#the-class-constructor)
		- [The use of  __ init __](#the-use-of---init)
		- [The 'self' keyword](#the-self-keyword)
	- [Class methods](#class-methods)
- [Leading and Trailing Underscores](#leading-and-trailing-underscores)
	- [Single leading underscore](#single-leading-underscore)
	- [Single trailing underscore](#single-trailing-underscore)
	- [Double leading underscore](#double-leading-underscore)
- [Tutorial](#tutorial)
- [Exercises](#exercises)

<!--TOC_END-->
## Overview

In this module we will cover some of the basics of OOP, mainly focussing on `classes` and `attributes`.

## Introduction to OOP

**O**bject **O**riented **P**rogramming has many principles (which will be discussed in a later module) that allow us to define complex types, known as `classes`. 
From these `classes` we create `objects` on which we can operate.

- `class` -  a set of *attributes* and *methods* which we can use to create objects.
- `object` - a usable instance of a class.

Since Python is an object oriented language, we can make use of these.

## Classes

> As we have discussed, a class is a set of `attributes` and `methods` whcih we can use to create `objects`.

Reviewer: ```suggestion
As we have discussed, a class is a set of `attributes` and `methods` which we can use to create `objects`.
```

- `attribute` - data, such as storing the height of something. 

> - `mehtod` - code, such as printing some text.

Reviewer: ```suggestion
- `method` - code, such as printing some text.
```

We would also need to define *functions* to replace the *methods* defined in the class.

All this is hard to keep track of and produces lots of duplicated code, which is `bad practice`.

### Class and Object example

We will continue with the Dog example from before, but this time see how we would define the `class` and `objects` in Python.

```python
class Dog:
    energy = "high"

    def speak(self):
        print("woof")

bilbo_waggins = Dog()
print(bilbo_waggins.energy)
bilbo_waggins.speak()
```

This defines the class `Dog`, sets the attribute `energy` equal to `"high"`, and defines the method `speak` to print `"woof"`.

Then sets the object `bilbo_waggins` to be a `Dog`.
Printing bilbo_waggins' energy will print `"high"`, and calling the method `speak()` will print `"woof"`.

We can redefine the attribute if our Dog doesn't have high energy.

```python
jk_growling = Dog()
jk_growling.energy = "low"
print(jk_growling.energy)
jk_growling.speak()
```

This will print the energy of `jk_growling` which we set to be `"low"`, and again calls the method `speak()`.

### The Class Constructor

> The class constructor allows us to set attributes when an obejct is created.

Reviewer: ```suggestion
The class constructor allows us to set attributes when an object is created.
```

Notice when we defined the class `Dog` we did not use parenthesis.
However when we defined the objects, `jk_growling = Dog()`, we did.

The method which we are calling here is the `constructor`, it is the method which is responsible for creating the instance of the class.

In our example it doesn't do a lot, but it can!

#### The use of  __ init __

This is a built in function, used with classes, which runs every time an object is being constructed.

```python
class Dog:
    def __init__(self, name, breed, energy):
        self.name = name
        self.breed = breed
        self.energy = energy
```

> Every time we construct a new object of this class, we must pass in a `name`, `breed` and `energy` parametersis a reference to the current instance of the class.

Reviewer: ```suggestion
Every time we construct a new object of this class, we must pass in `name`, `breed` and `energy` parameters as a reference to the current instance of the class.
```

Reviewer: Is this what you are trying to say?

```python
dog1 = Dog("Ross Barkley", "Jack Russel", "High")
```

#### The 'self' keyword

The `self` parameter is a reference to the current instance of the class.

`Class data` members are usually set as default and belong to the class, whereas `instance data` members are unique to the instance of the class.

We need a way to map one to the other, and that's what the `self` parameter does.

Note, the paramter does not need to be called `self`, that is just best practice, but it **must** be the `first` parameter of any class function!

### Class methods

There are some useful methods to know:

- `getattr(object, name)` - Accesses the named attribute of a specific object.
- `hasattr(object, name)` - Check if a specific object has a named attribute.
- `setattr(object, name, value)` - Sets the named attribute of a specific object as the value specified.
- `delattr(object, name)` - Deletes the named attribute of a specific object.

## Leading and Trailing Underscores

### Single leading underscore

Used to signify a *private* or *internal* variable, these variables should not be accessed.

> For example, `_money = 1,000,000` as we someone's balance is private.

Reviewer: ```suggestion
For example, `_money = 1,000,000` as someone's balance is private.
```

### Single trailing underscore

Used to avoid conflicts with Python *keywords*.

For example, we might define `class_ = "QAC Intake"`, but as we know `class` is a *keyword* in Python so we must avoid conflicting with it.

### Double leading underscore

A double leading underscore applied `name mangling` to the class.

> This is where the interpreter chnages the name of the variable to make it harder to create collisions when the class is extended.

Reviewer: ```suggestion
This is where the interpreter changes the name of the variable to make it harder to create collisions when the class is extended.
```

We access variables which have double leading underscores in the following way:
`_ClassName__VariableName`.

```python
class Feline:
    __family = "Felidae"

kitty = Feline()
print(kitty._Feline__family)
```

> This would print `"Felidae"`, the value of the `__family` varibale.

Reviewer: ```suggestion
This would print `"Felidae"`, the value of the `__family` variable.
Whereas `print(kitty.__family)` wouldn't print anything.
```
